src/variorum/ARM/config_arm.c

```diff
@@ -8,34 +8,63 @@
 
 #include <config_arm.h>
 #include <config_architecture.h>
-#include <power_features.h>
-#include <ARM.h>
+#include "power_features_Juno_r2.h"
+#include "power_features_Neoverse_N1.h"
+#include <ARM_Juno_r2.h>
+#include <ARM_Neoverse_N1.h>
 #include <variorum_error.h>
 
 uint64_t *detect_arm_arch(void)
 {
     // ARM systems come in various flavors. For now, we will use a default model number.
-    uint64_t *model = (uint64_t *) malloc(sizeof(uint64_t));
-    *model = ARMV8;
+    unsigned long *model = (unsigned long *) malloc(sizeof(uint64_t));
+    asm volatile(
```

Comment:
I haven't tested this change, @amarathe84, I assume you have thoroughly tested the part where it picks up the model. Is there documentation on this that we can add somewhere?

Also, I noticed that for Juno r2, you now have a different model for the big and the little processors, as opposed to just one which we had before. Can you elaborate why? Should we be representing them as the same model as we have been viewing the big.Little as a single entity?

Reply:
I updated the PR description with the integration/regression tests on nhvpc1 and Juno r2. Please take a look at the test outcomes to see if they look okay. 

Let me also post the description for the updated model ID check here shortly.

Reply:
In the existing ARM implementation we assumed a generic ARMV8 (constant) model. To distinguish between ARM implementations, we need to look at the 'Primary part number' [15:4] bit fields of the MIDR_EL1 (Main ID) register which is defined per ARM CPU implementation (as opposed to a combined SoC like Juno r2). Here are the links to the MIDR_EL1 register for the three ARM CPU architectures we support:

Cortex A53:
https://developer.arm.com/documentation/ddi0500/j/System-Control/AArch64-register-descriptions/Main-ID-Register--EL1

Cortex A72:
https://developer.arm.com/documentation/100095/0003/System-Control/AArch64-register-descriptions/Main-ID-Register--EL1?lang=en

Neoverse N1:
https://developer.arm.com/documentation/100616/0301/register-descriptions/aarch64-system-registers/midr-el1--main-id-register--el1

Based on the model ID we set up the lower-level interfaces.

Juno r2 is a big.LITTLE implementation with both Cortex A72 and Cortex A53 in a single SoC. There are systems (e.g. revisions of Raspberry Pi) with either one of these two but with the same interfaces to lower-level functionality so the same code should work with them, but we haven't tested on such systems. 

There's also a filesystem interface for MIDR_EL1 but I couldn't confirm if that's always available on an Arm implementation, so I went with the MRS instruction to get the model ID instead. 

I'll add a subsection in the ARM Overview about model identification along with the links to ARM documentation.

Reply:
Thank you for explaining @amarathe84 ! This is very helpful for me to understand the detail of the model/arch_id. And yes, this would be great to document somewhere outside of this issue too in the ARM documentation in some way. I didn't know about the MIDR_EL1  register.